### ISSUE TITLE
MDCACHE LRU reaper thread issue

### ISSUE THREAD

**Issue by deepakaru** [nfs-ganesha/nfs-ganesha#963]
We recently hit some issues where the number of ganesha mdcache entries grew to 6-9 Million when the mdcache hiwater mark was at 450,000.

### ROOT CAUSE

Upon further debugging we found that readdir was not playing well with garbage collection. The major issue comes from the fact that we have a temporary pointer from the dirent to the inode mdcache entry.

Today when readdir fills up a chunk it puts a temporary reference from dirent to the inode cache entry. This happens inside 
```
mdc_readdir_chunk_object
            ...
            new_dir_entry->entry = new_entry;

```
Then we clear it out inside,

```
mdcache_readdir_chunked 
      ...  
           if (has_write && dirent->entry) {
      ...
                  mdcache_lru_unref(dirent->entry, LRU_FLAG_NONE);
                  dirent->entry = NULL;
            }
          ...
```
But this means each readdir can hold a refcount on 128 entries in the inode cache at a time.

This leads to a scenario where the following happens
1. We have multiple readdirs going in parallel each reading different directories with 1 Million entries. Each readdir ends up holding a refcount of 2 on 128 entries - some of which end up being the entries at the LRU edge. Because of this the reaper thread is not able to find an entry to reap
2. There is also another bug where we are not hitting the clearing out code path in mdcache_readdir_chunked. We don't have an RCA on where this bug is - but we often see systems where all the LRU edges of the inode cache queues are filled with a refcount of 2 even when there is no load on the system. And all the entries are being referenced by their dirents which is the cause of the refcount.

### 1. Can we remove this temporary pointer from dirent to the normal mdcache entry? 

This optimization is only helpful in the cache miss code path. I think this optimization accomplishes two things,

- We don't have to do a lookup from dirent to inode entry in the readdir cache miss code path. But since we are already in the readdircache miss code path I don't think the performance penalty will even be visible.
- It prevents the inode cache entries from being flushed out when a readdir is in progress. I wonder if we should just use the LRU way to accomplish this - which brings me to the second question

### 2. Can we always move an entry to the MRU of L2 even during scans? 

Today this is a bit of an inconsistent behavior - we insert new entries created during readdir at the MRU of L2 but we don't adjust old entries to the MRU. Not sure if this is an intended behaviour or bug. The following is a detailed explanation of this behaviour,

 

Inside mdc_readdir_chunk_object we call mdcache_new_entry with LRU_FLAG_NONE

```
 status = mdcache_new_entry(export, sub_handle, attrs_in, false, NULL,
                           false, &new_entry, NULL, LRU_FLAG_NONE);

```
if its a inode cache hit it follows this path

```
status = mdcache_find_keyed_reason(&key, entry, flags);
....        
     if (likely(*entry)) {
            fsal_status_t status;
...
            /* Initial Ref on entry */
            mdcache_lru_ref(*entry, flags);    
...
 
```
since flags is not LRU_REQ_INITIAL this entry doesn't get adjusted to the MRU of L2 or LRU of L1
but if its a inode cache miss it follows this path
`   ....            mdcache_lru_insert(nentry, flags);`

 if flags has LRU_FLAG_NONE mdcache_lru_insert inserts the entry into the MRU of L2

**Comment by Steve-Lord** [comment]
The LRU cache logic has some bad behavior in this area. inode entries from directory reads can end up at the tail of all the lanes in the LRU. Once this happens, the code to attempt to reclaim an entry for another inode fails to make progress and it just allocates a new one.  I have seen 23 million entries in the inode cache with a high water mark of 2 million.

The reaper thread has variable aggressiveness, depending on the open fd count, if files are not open, it runs at its default rate and reclaims very slowly.

I added a couple more counters to the lru stats and generated this state:

```
LRU Utilization Data
 FSAL opened FD count :                     0
 System limit on FDs :                1048576
 FD usage :               Below Low Water Mark
 LRU entries in use :                 5620868
 LRU entries reclaim :               15988023
 LRU entries busy :                 120837851
 Chunks in use :                        25174
```

reclaim is cases where it managed to reclaim a cache entry from the LRU, busy is where it failed to do so because of reference counts.

A workload which generates content and then starts scanning with readdir tends to get into this state when performed at scale. Unless you run out of memory it tends not to be a visible problem.

The reference counts from the directory chunk cache are preventing lru_reap_impl() call from keeping the cache within configured bounds. Going to experiment with reclaiming from the chunk cache when the high water mark has been reached, have not read enough code yet to understand the benefits of the dir chunk cache cross referencing the LRU cache.

**Comment by deepakaru** [comment]
I think reaping from the chunk LRU might not be a complete solution to the problem. I am sure we can come up with case where this solution works but there are also corner cases where it might not.

Consider the following scenario, A is a chunk LRU entry referencing B which is a inode LRU entry
A->B
Lets say B is at the end of the inode LRU queue,
- if A is at the tail or 'near the tail' of the chunk LRU queue what you are proposing might work
- but if A is at the head of the chunk LRU, we can just clear out the entire chunk LRU queue
From my understanding of these areas, the above scenario is very much possible

Thats why i am proposing 
1. Can we remove this temporary pointer from dirent to the normal mdcache entry?
2. Can we always move an entry to the MRU of L2 even during scans?

**Comment by mattbenjamin** [comment]
I think moving the entry to MRU *of L2* preserves scan resistance, so if this isn't a typo, I could support it.  (if the entry is already in L1, ofc, we shouldn't demote it)

**Comment by deepakaru** [comment]
thanks @mattbenjamin . that isn't a typo. We don't move existing entries to the MRU of L2 today

**Comment by ffilz** [comment]
We had a conversation on the Ganesha community call this morning and we came up with some thoughts:

First, the number of entries the chunk cache implicates (i.e. number of chunks * number of entries in a chunk) should be no greater than the entry cache hi-water mark. This will assure that the dirent cache doesn't easily lock up all the entry cache entries.

Second, the chunk cache will take a long term reference on the entries referenced by the dirent cache.. This removes them from LRU L1 or L2. This represents that until the chunk is reaped, those entries are locked in and not a candidate for reaping.

Third, we will track for entries if they have ever been referenced for purposes other than directory scan. If not, when the chunk is reaped and the long term references are dropped, the entry will be placed at the MRU of L2 instead of the MRU of L1. This will be accomplished by clever use of the flags for mdcache_lru_unref (when readdir releases the LRU_LONG_TERM_REFERENCE it will do just that, while mdcache_put_long_term_ref (called by all the existing long term ref users) will also use the LRU_PROMOTE flag).

We need to add reaping of chunks if the chunk case is above the hiwater mark (currently chunk_lru_run_lane ONLY demotes from L1 to L2...). We should also reap chunks on an age basis all the way down to 0 (or add a configurable low water mark).

I'm not sure if L2 makes sense for the chunk LRU...

**Comment by dang** [comment]
This sounds good.  I vote for a low-water mark for chunks, default to something small (10%?  20%?).  I don't know if L2 makes sense, but it also doesn't hurt, so I'd say leave it for now, unless it's causing' problems.

**Comment by ffilz** [comment]
I've coded something up in this patch and the two related patches (but it doesn't yet include chunk reaping... oops...):

https://review.gerrithub.io/c/ffilz/nfs-ganesha/+/556252/3

Looking at setting a limit, the defaults would only allow about 700 chunks for 100,000 entries. That actually seems pretty small. But with pretty quick chunk reaping, it should also allow pretty quick cache reaping. This aspect needs more thought.

**Comment by Steve-Lord** [comment]
Thanks,

I pulled this, threw it onto a system and fired up a test load and  let it run overnight

I have explicit limits right now:

    "MDCACHE": {
        "NParts": "5039",
        "Cache_Size": "48017",
        "Entries_HWMark":"4000000",
        "Chunks_HWMark": "25000",
        "Reaper_Work_Per_Lane": 1000
    },


If I read it aright, this would keep the entries from the chunk cache at no more than 80% of the entry cache. I probably want leave more room outside the chunk cache now that I look at that.

What I  noticed initially with this change was that the chunk cache appears to be getting cannibalized, previously it would tend towards its configured limit and stay around there. Now it can get pushed down very small and seems to be getting reclaimed much more aggressively. The entry cache is still pushing past the configured limit. This eventually broke down and the cache size started to grow more and more aggressively.

Having let the system run for about 5 hours of increasing load I ended up back in the same corner I did before:

Inode Cache statistics
 Cache Requests:                            0
 Cache Hits:                        798095445
 Cache Misses:                      480697344
 Cache Conflicts:                           0
 Cache Adds:                        358328001
 Cache Mapping:                     716656002

LRU Utilization Data
 FSAL opened FD count :                     0
 System limit on FDs :                1048576
 FD usage :               Below Low Water Mark
 LRU entries in use :                21152376
 Chunks in use :                        24871

Looking at the breakdown of content inside the cache:

Lane 0 L2  size 167012
Lane 0 L1  size 1076905
Lane 0 long_term  size 0
Lane 1 L2  size 167012
Lane 1 L1  size 1077178
Lane 1 long_term  size 1
Lane 2 L2  size 167012
Lane 2 L1  size 1076978
Lane 2 long_term  size 0
Lane 3 L2  size 167012
Lane 3 L1  size 1077126
Lane 3 long_term  size 0
Lane 4 L2  size 167012
Lane 4 L1  size 1076905
Lane 4 long_term  size 0
Lane 5 L2  size 167011
Lane 5 L1  size 1077208
Lane 5 long_term  size 0
Lane 6 L2  size 167011
Lane 6 L1  size 1077527
Lane 6 long_term  size 0
Lane 7 L2  size 167011
Lane 7 L1  size 1076904
Lane 7 long_term  size 0
Lane 8 L2  size 167011
Lane 8 L1  size 1077164
Lane 8 long_term  size 0
Lane 9 L2  size 167012
Lane 9 L1  size 1077581
Lane 9 long_term  size 0
Lane 10 L2  size 167012
Lane 10 L1  size 1077074
Lane 10 long_term  size 0
Lane 11 L2  size 167012
Lane 11 L1  size 1076932
Lane 11 long_term  size 0
Lane 12 L2  size 167012
Lane 12 L1  size 1077136
Lane 12 long_term  size 1
Lane 13 L2  size 167012
Lane 13 L1  size 1077108
Lane 13 long_term  size 0
Lane 14 L2  size 167012
Lane 14 L1  size 1077067
Lane 14 long_term  size 0
Lane 15 L2  size 167012
Lane 15 L1  size 1077220
Lane 15 long_term  size 1
Lane 16 L2  size 167012
Lane 16 L1  size 1077160
Lane 16 long_term  size 0

Long term appears not to be used at all here.

I suspect that under enough pressure lru_reap_impl Is always finding entries with a high ref count at the head of the lanes. Consider two threads competing on the same lane. One obtains the lock, bumps the ref, drops the lock, the second thread obtains the lock while first is in cih_latch_entry and we not have an entry which neither thread can use. They then move on to next lane. I tried a version which keeps walking down the list until it finds an entry which does not have a raised ref count. This still  breaks down, but it keeps the cache size under control for longer. Not recommending that as a solution, seems more like something which removes the entry under the first QLOCK and puts it back if it is rejected would avoid contention here. I think this is a bit of a stretch as a cause, I only have 96 rpc service threads (64 cores)and there is a lot of other things they will be doing. I have run an instrumented version of the code and the high ref is, with a very few exceptions why the lane is not reclaimed from. reap_lane is 3214976494 after this, so there have been a lot of iterations around the lanes.

Going to try and find other evidence for this one way or the other. Is there any other place other than pulling off the lru that the first entry should have a high ref count?

Having said all that I realized I only applied the one patch, and I should have applied all three. 

Steve

**Comment by ffilz** [comment]
Steve, I'm not sure what has changed the reaping of chunks because there is currently no code to reap them on a continuing basis, so they will only get reaped by a new READDIR. Since they aren't getting reaped, the mdcache entries associated with them will not have their refcount go to 1.

I'm not sure why you aren't seeing elevated long term ref count

**Comment by ffilz** [comment]
Additional thinking about the chunk cache.

One option would be to consider the utilization of the inode cache in driving pressure on the chunk cache.

We probably should also change the default chunk cache high water mark, reduce it by a factor of 10 or even 100. We could document that when configuring for a workload with lots of very small directories (say average size of N), the value should be set to entries_hwmark / N.

**Comment by Steve-Lord** [comment]
I suspect the chunks are being recycled based on readdirs here, I was just unexpected.

A little disclosure, the ‘workload’ is the spec 2020 swbuild benchmark. It creates a large amount of small file content and then runs a workload intended to similate compilation in a large filesystem. It runs many instances of this in parallel. I am just using this to generate large file count load.

I have isolated a subcase in the test where there are no readdirs and nothing in the L2 lanes at all. If I run the test with existing content in place when its working set is large (say 48 million files), then during its initial phase it is basically discovering its content exists:

Duration: 48.1282672882 seconds

Operation Details                            |  Operation Latency (in milliseconds)
=============================================|========================================
Name                        Total     Error  |       Avg          Min           Max
ACCESS                      20750         0  |     0.000831     0.000230     0.054329
CLOSE                           6         0  |     0.006271     0.004090     0.009709
CREATE                      20535     20532  |     0.145923     0.025679     1.852079
GETATTR                   1083843         0  |     0.021872     0.000630     3.051858
GETFH                     1062900         0  |     0.000571     0.000220     0.118618
LOOKUP                    1062894         1  |     0.106625     0.001130     8.069308
OPEN                            6         0  |     0.215507     0.014990     0.323855
PUTFH                     1104386         0  |     0.004356     0.001990     1.097611
READDIR                         1         0  |     0.234906     0.234906     0.234906
REMOVE                          7         0  |     0.335383     0.292055     0.366394
SETATTR                         1         0  |     0.170897     0.170897     0.170897
WRITE                           2         0  |     0.565101     0.509981     0.620220
SEQUENCE                  1104387         0  |     0.001308     0.000410     0.113838

Those CREATEs are mkdirs on existing directories.

While this is ongoing, the cache is growing rapidly  past its hwmark and we are finding L1 cache entries we cannot use.

Because this only happens under heavy load it is hard to catch why the ref counts are not sitting at 1, if I grab it with gdb they always look like they could be reclaimed.

Steve

**Comment by ffilz** [comment]
I removed the patch limiting Chunks_HWMark but added a patch to implement reaping and add Chunks_LWMark,

https://review.gerrithub.io/c/ffilz/nfs-ganesha/+/556283

I think that patch needs a bit more consideration, but I wanted to get it out there for discussion and testing. Play with the tunables and see what makes sense.

I suspect real use cases will want to carefully pick Chunks_HWMark and Chunks_LWMark.

We may also want a different lru_run_interval for the chunk cache.

**Comment by ffilz** [comment]
OK, another bit on this... The dirent should not actually have a long term reference. It's not supposed to be long term, but somehow it's sticking as long term. I need to look at the code and figure out what happened...

Hmm, but actually, maybe the long term ref is still good, as long as once the readdir is done with the chunk that the long term reference goes away. If that actually is the case, then the long term ref accomplishes option 2 from the OP of move to MRU, except a bit better, we remove during the operation, and place at MRU of L2 at the conclusion.

But there may still be a problem with too many concurrent readdirs...

One solution to that would be smaller chunks if the number of concurrent readdirs * the size of a chunk implicates too many inodes.

More thoughts and analysis coming,,,

**Comment by Steve-Lord** [comment]
Except I can induce uncontrolled cache growth without a single readdir being performed at all. I suspect there is more than one problem here…. This is with the second patch set by the way.

There is a reader / writer lock on the trees used to lookup keys. A read lock across the lookup of an entry and a write lock when inserting / removing one. The tree chosen is based on the key of the record, which is about the only common sense thing to use. The lookup path does:

	Lock tree for read
	Find entry, return with tree locked

        Call mdcache_lru_ref()

		Increment ref count
		
	       adjust lru position IF AND ONLY IF (long term reference or REQ_INITIAL)

        Drop the tree lock.

There are several calls where the passed in flags are LRU_FLAG_NONE and an entry will not move on the LRU.

On the reclaim path it grabs the QLOCK, increments the reference and drops the QLOCK. Then it goes after the tree lock for write for the entry at the tail of the LRU.

There are some keys which are likely hot in the cache, the root, various directories etc. The use of a read lock in one path and a write lock in the other could cause the writer to have to wait for several readers before proceeding. Combine that with a hot directory which is often referenced. I do not have a good feel for the duration of a reference, but if it is across calls to the underlying FSAL it could be measurable. That could lead to conditions where an entry with an elevated ref count sits at the end of a lane and is repeatedly rejected due to the the ref count.

In theory it moves on to another lane, and finds another candidate. I added some tracing to where the lru_reap_impl was rejecting entries - the ref count is 3, sometimes 4 and this is always the first check on the refcount. They were all on the same lane for a while - thousands of them in sequence, then they moved to another lane. This was coming from many threads at once.

I do wonder about this code from lru_reap_impl:

	lane = LRU_NEXT(reap_lane);
	for (ix = 0; ix < LRU_N_Q_LANES; ++ix, lane = LRU_NEXT(reap_lane)) {

If this fails to find anything it ends up back where it started. I was running with this code in its place and the cache was hardly growing past the cap at all. Unfortunately we just lost power in our lab and I think I am done for a while.

		lane = LRU_NEXT(reap_lane);
	for (ix = 0; ix < LRU_N_Q_LANES; ++ix, lane++) {
	    if (lane >= LRU_N_Q_LANES)
	        lane = 0;


I am still not sure what the root cause is here, but this was a noticeable change in behavior. I did not get a chance to look at the impact on the balance between lanes before the power went out though.

Steve

**Comment by deepakaru** [comment]
We found another issue in this readdir - mdcache code path where we were not handling an error properly. Have a fix for that issue here https://review.gerrithub.io/c/ffilz/nfs-ganesha/+/556455

**Comment by Steve-Lord** [comment]
Thanks, 

Will add this to the collection, however, even with Dir_Chunk set to zero and the cache disabled I can still drive the cache well past its high water mark. I have spent this week working on other issues, hope to make it back to this next week. I have been unable to find where the ref count is coming from in my case, but it is transient in that I can subsequently look with gdb and find the ref counts low again. There must be some code path which holds the count high for long enough that the reclaim fails frequently.

Steve

**Comment by deepakaru** [comment]
@Steve-Lord , 

i agree - i've seen something similar in my tests as well - i've been running nfstest_io.

I think we have to bugs here. One is the readdir code path - which this fix https://review.gerrithub.io/c/ffilz/nfs-ganesha/+/556455 and other fixes from @ffilz are addressing.

But even with these fixes i see scenarios where we are not able to reap entries from the cache. I know from some instrumentation i've added that these unreapable entries are not from readdir. This must be another bug in some other code path where we are not releasing the refcount properly.

I am trying to figure out what op these are coming from as well

**Comment by Steve-Lord** [comment]
Not sure it is a leak, it feels more like there is a case where this entry is held referenced for long enough that the cache cannot use any of the entries, when I walk the lists in gdb on an idle system all the ref counts are good again.
In my case the first test on the ref count rejects it, Heavy contention on the cih_latch_entry call could cause this, but that should not be hot - unless the hashing of keys to partitions is very unfair and there is a readlock out which is blocking the reclaims.

Steve

**Comment by ffilz** [comment]
Hmm, could you try with a lower cap on the readdir cache? You could also try a lower low water mark on the chunk cache so we drive out the mdcache entries associated with readdir faster.

If the entries that are part of readdir aren't being used for regular access, they would live in L2 and not block reclaim from L1. If that's the case, then if the cache is growing above the high water mark, it would imply that all the entries in L1 have elevated refcount and therefore the working set is larger than the cache size.

EXCEPT... we still have a problem, and maybe you're hitting this. If the LRU of L1 happens to be used in a long operation, then it may have an elevated refcount longer than an entry that was more recently referenced but for a short time.

This IS speaking to the thought that maybe all entries with elevated refcount need to be removed from the LRU. But there was some concern then about the order of returning no longer referenced entries back into the LRU.

**Comment by ffilz** [comment]
I just had a conversation with Matt Benjamin. We came the the conclusion that we really need to get all the elevated refcounts out of the LRU. I'm going to post a patch making them long term references. If that ends up the solution, we will rename from long term reference to active reference.

I'm not quite sure but there might be some non-active references other than the sentinel reference for being reachable in the cache, but I suspect we will basically want ANY elevated reference be an active reference.,

**Comment by ffilz** [comment]
Try out this patch https://review.gerrithub.io/c/ffilz/nfs-ganesha/+/556554?usp=search

**Comment by ffilz** [comment]
There may still be some more work to do. Someone else is working on a patch to allow the number of LRU lanes to be configurable since that may be an issue also.

**Comment by Steve-Lord** [comment]
Thanks, will take it for a spin, probably tomorrow now. On the more lanes idea, it is a bit of a double edged sword, you can spend longer running around attempting to find a candidate if there are more of them. 


Steve

p.s. I am not a regular user of Gerrit, I presume there is an order to apply these in and is the ’Submitted together’ set that should be applied?

**Comment by ffilz** [comment]
The relation chain shows the patches in order, with the top/1st one being "HEAD". If you open that patch, there is a "checkout" option in the download drop down. I usually paste that and then add "-b branch-name" before hitting enter. That will check out the whole set of patches as a branch in git.

**Comment by Steve-Lord** [comment]
This code looks wrong:

static inline void
make_inactive_lru(mdcache_entry_t *entry)
{
    mdcache_lru_t *lru = &entry->lru;
    struct lru_q_lane *qlane = &LRU[lru->lane];
    struct lru_q *q;

    switch (lru->qid) {
    case LRU_ENTRY_L1:
    case LRU_ENTRY_L2:
        assert(false);
        break;
    case LRU_ENTRY_ACTIVE:
        /* Move entry to MRU of L1 or L2 */
        q = lru_queue_of(entry);
        LRU_DQ(&entry->lru, q);

        if (atomic_fetch_uint32_t(&entry->lru.flags)
            & LRU_EVER_PROMOTED) {
            /* If entry was ever promoted, insert into L1. */
            q = &qlane->L1;
        } else {
            /* Entry was never promoted, only ever used in a
             * directory scan, return to L2.
             */
	    q = &qlane->L1; <—— L2 Here?
        }

        lru_insert(&entry->lru, q);
        break;
    case LRU_ENTRY_CLEANUP:
    default:
        /* do nothing */
        break;
    }   /* switch qid */
}


That said now eveything moved to the active list all the time, so the L1 and L2 lists are empty and the cache still grows, possibly a missing inactive somewhere. I still need to read the diff more. I am looking into getting reference count tracing working on top of this patch. I find I want to know more than mdcache_get_ref and mdcache_put_ref which are the bulk of it right now. It is the callers the other end of the FSAL function vector which are interesting in this case.

I am juggling this with other work at the moment so be patient…

Steve

**Comment by ffilz** [comment]
Hmm, maybe there's a refcount issue somewhere.

I fixed that L1 that should have been L2...

**Comment by Haroldll** [comment]
I'm also bothered by this issue, entries_used often far exceeds entries_hiwat. Dir_Chunk is 0, all other MDCACHE parameters are configured by default. As long as the client's business continues to be busy, the lru_try_reap_entry basically returns NULL, unable to recycle entry. Ganetha uses too much memory and often OOM...
So, lru_try_reap_entry couldn't succeed, for what? Why refcnt has not been 2?
Is there any configuration to avoid this problem?
Thanks!

**Comment by ffilz** [comment]
> I'm also bothered by this issue, entries_used often far exceeds entries_hiwat. Dir_Chunk is 0, all other MDCACHE parameters are configured by default. As long as the client's business continues to be busy, the lru_try_reap_entry basically returns NULL, unable to recycle entry. Ganetha uses too much memory and often OOM... So, lru_try_reap_entry couldn't succeed, for what? Why refcnt has not been 2? Is there any configuration to avoid this problem? Thanks!

Can you try with my patch set as described? There may be a leaked refcount somewhere but I'm not quite sure how to find it.

**Comment by Haroldll** [comment]
> 
> Can you try with my patch set as described? There may be a leaked refcount somewhere but I'm not quite sure how to find it.

I tried this patch: https://review.gerrithub.io/c/ffilz/nfs-ganesha/+/556554, but the problem doesn't seem to be solved.

**Comment by Steve-Lord** [comment]
I have been experimenting with tracing of the changes to the active_rrefcnt and where the call came from:

Entry 0x7f0146803000 refcnt: 1 maxidx 30
 idx 0 add 1 refcnt 2 mdcache_alloc_and_check_handle + 100 in section .text of /lib/libganesha_nfsd.so.5.3.2
 idx 1 add 1 refcnt 3 _state_add_impl + 875 in section .text of /lib/libganesha_nfsd.so.5.3.2
 idx 2 add 1 refcnt 4 open4_create_fh + 137 in section .text of /lib/libganesha_nfsd.so.5.3.2
 idx 3 add -1 refcnt 3 open4_create_fh + 155 in section .text of /lib/libganesha_nfsd.so.5.3.2
 idx 4 add -1 refcnt 2 compound_data_Free + 98 in section .text of /lib/libganesha_nfsd.so.5.3.2

 idx 5 add 1 refcnt 3 mdcache_locate_host + 373 in section .text of /lib/libganesha_nfsd.so.5.3.2
 idx 6 add 1 refcnt 4 nfs4_mds_putfh + 706 in section .text of /lib/libganesha_nfsd.so.5.3.2
 idx 7 add -1 refcnt 3 nfs4_mds_putfh + 729 in section .text of /lib/libganesha_nfsd.so.5.3.2
 idx 8 add -1 refcnt 2 compound_data_Free + 98 in section .text of /lib/libganesha_nfsd.so.5.3.2

 idx 9 add 1 refcnt 3 mdcache_locate_host + 373 in section .text of /lib/libganesha_nfsd.so.5.3.2
 idx 10 add 1 refcnt 4 nfs4_mds_putfh + 706 in section .text of /lib/libganesha_nfsd.so.5.3.2
 idx 11 add -1 refcnt 3 nfs4_mds_putfh + 729 in section .text of /lib/libganesha_nfsd.so.5.3.2
 idx 12 add 1 refcnt 4 get_state_obj_export_owner_refs + 127 in section .text of /lib/libganesha_nfsd.so.5.3.2
 idx 13 add -1 refcnt 3 nfs4_Check_Stateid + 1129 in section .text of /lib/libganesha_nfsd.so.5.3.2
 idx 14 add -1 refcnt 2 compound_data_Free + 98 in section .text of /lib/libganesha_nfsd.so.5.3.2

 idx 15 add 1 refcnt 3 mdcache_locate_host + 373 in section .text of /lib/libganesha_nfsd.so.5.3.2
 idx 16 add 1 refcnt 4 nfs4_mds_putfh + 706 in section .text of /lib/libganesha_nfsd.so.5.3.2
 idx 17 add -1 refcnt 3 nfs4_mds_putfh + 729 in section .text of /lib/libganesha_nfsd.so.5.3.2
 idx 18 add -1 refcnt 2 compound_data_Free + 98 in section .text of /lib/libganesha_nfsd.so.5.3.2

 idx 19 add 1 refcnt 3 mdcache_locate_host + 373 in section .text of /lib/libganesha_nfsd.so.5.3.2
 idx 20 add 1 refcnt 4 nfs4_mds_putfh + 706 in section .text of /lib/libganesha_nfsd.so.5.3.2
 idx 21 add -1 refcnt 3 nfs4_mds_putfh + 729 in section .text of /lib/libganesha_nfsd.so.5.3.2
 idx 22 add 1 refcnt 4 get_state_obj_export_owner_refs + 127 in section .text of /lib/libganesha_nfsd.so.5.3.2
 idx 23 add -1 refcnt 3 nfs4_Check_Stateid + 1129 in section .text of /lib/libganesha_nfsd.so.5.3.2
 idx 24 add 1 refcnt 4 get_state_obj_export_owner_refs + 127 in section .text of /lib/libganesha_nfsd.so.5.3.2
 idx 25 add 1 refcnt 5 _state_del_locked + 165 in section .text of /lib/libganesha_nfsd.so.5.3.2
 idx 26 add -1 refcnt 4 _state_del_locked + 533 in section .text of /lib/libganesha_nfsd.so.5.3.2
 idx 27 add -1 refcnt 3 _state_del_locked + 604 in section .text of /lib/libganesha_nfsd.so.5.3.2
 idx 28 add -1 refcnt 2 nfs4_op_close + 721 in section .text of /lib/libganesha_nfsd.so.5.3.2
 idx 29 add -1 refcnt 1 compound_data_Free + 98 in section .text of /lib/libganesha_nfsd.so.5.3.2

A circular buffer of 32 entries which traces if it was an inc or a dec the ref count after the operation and the return address from the call to the ->put ->get operations in mdcache. These are under a global lock so are correctly sequenced. I then have some gdbmacros, I run a small load and then use the gdbmacros to dump the cache contents. Here we see an entry which has a current ref count of 1 on the active lanes. I added spacing here around what appear to be the different RPC operations on the entry. 

Here we have, I think, create to close on a file. There is clearly a reference held across this whole sequence, but it feels like there is an extra ref at the end of create itself which leaves us one high at the end of close. 

Not sure if this helps or not. Patch on top of the active lru change plus gdb macros attached, not recommended for anything debugging. These reference counts are pretty hot.

Steve

￼

**Comment by ffilz** [comment]
Could you track both the main refcount and the active refcount? This sounds like a situation where the main refcount was released but not the active, so a missing flag on the unref.

**Comment by Steve-Lord** [comment]
This is capturing the value of refcnt at the same trace points as active_refcnt, not sure it tells you anything beyond the two values
Tracking each other with the refcnt being one more than the active refcnt. If you meant trace refcnt when it is changing that
would be a more substantial change I expect.

Entry 0x7f0e2acef000 refcnt: 2 actref: 1 maxidx 30
 idx 0 add 1 refcnt 3 actref 2 mdcache_alloc_and_check_handle + 100 in section .text of /lib/libganesha_nfsd.so.5.3.2
 idx 1 add 1 refcnt 4 actref 3 _state_add_impl + 875 in section .text of /lib/libganesha_nfsd.so.5.3.2
 idx 2 add 1 refcnt 5 actref 4 open4_create_fh + 137 in section .text of /lib/libganesha_nfsd.so.5.3.2
 idx 3 add -1 refcnt 4 actref 3 open4_create_fh + 155 in section .text of /lib/libganesha_nfsd.so.5.3.2
 idx 4 add -1 refcnt 3 actref 2 compound_data_Free + 98 in section .text of /lib/libganesha_nfsd.so.5.3.2

 idx 5 add 1 refcnt 4 actref 3 mdcache_locate_host + 373 in section .text of /lib/libganesha_nfsd.so.5.3.2
 idx 6 add 1 refcnt 5 actref 4 nfs4_mds_putfh + 706 in section .text of /lib/libganesha_nfsd.so.5.3.2
 idx 7 add -1 refcnt 4 actref 3 nfs4_mds_putfh + 729 in section .text of /lib/libganesha_nfsd.so.5.3.2
 idx 8 add -1 refcnt 3 actref 2 compound_data_Free + 98 in section .text of /lib/libganesha_nfsd.so.5.3.2
 idx 9 add 1 refcnt 4 actref 3 mdcache_locate_host + 373 in section .text of /lib/libganesha_nfsd.so.5.3.2
 idx 10 add 1 refcnt 5 actref 4 nfs4_mds_putfh + 706 in section .text of /lib/libganesha_nfsd.so.5.3.2
 idx 11 add -1 refcnt 4 actref 3 nfs4_mds_putfh + 729 in section .text of /lib/libganesha_nfsd.so.5.3.2
 idx 12 add 1 refcnt 5 actref 4 get_state_obj_export_owner_refs + 127 in section .text of /lib/libganesha_nfsd.so.5.3.2
 idx 13 add -1 refcnt 4 actref 3 nfs4_Check_Stateid + 1129 in section .text of /lib/libganesha_nfsd.so.5.3.2
 idx 14 add -1 refcnt 3 actref 2 compound_data_Free + 98 in section .text of /lib/libganesha_nfsd.so.5.3.2

 idx 15 add 1 refcnt 4 actref 3 mdcache_locate_host + 373 in section .text of /lib/libganesha_nfsd.so.5.3.2
 idx 16 add 1 refcnt 5 actref 4 nfs4_mds_putfh + 706 in section .text of /lib/libganesha_nfsd.so.5.3.2
 idx 17 add -1 refcnt 4 actref 3 nfs4_mds_putfh + 729 in section .text of /lib/libganesha_nfsd.so.5.3.2
 idx 18 add -1 refcnt 3 actref 2 compound_data_Free + 98 in section .text of /lib/libganesha_nfsd.so.5.3.2

 idx 19 add 1 refcnt 4 actref 3 mdcache_locate_host + 373 in section .text of /lib/libganesha_nfsd.so.5.3.2
 idx 20 add 1 refcnt 5 actref 4 nfs4_mds_putfh + 706 in section .text of /lib/libganesha_nfsd.so.5.3.2
 idx 21 add -1 refcnt 4 actref 3 nfs4_mds_putfh + 729 in section .text of /lib/libganesha_nfsd.so.5.3.2
 idx 22 add 1 refcnt 5 actref 4 get_state_obj_export_owner_refs + 127 in section .text of /lib/libganesha_nfsd.so.5.3.2
 idx 23 add -1 refcnt 4 actref 3 nfs4_Check_Stateid + 1129 in section .text of /lib/libganesha_nfsd.so.5.3.2
 idx 24 add 1 refcnt 5 actref 4 get_state_obj_export_owner_refs + 127 in section .text of /lib/libganesha_nfsd.so.5.3.2
 idx 25 add 1 refcnt 6 actref 5 _state_del_locked + 165 in section .text of /lib/libganesha_nfsd.so.5.3.2
 idx 26 add -1 refcnt 5 actref 4 _state_del_locked + 533 in section .text of /lib/libganesha_nfsd.so.5.3.2
 idx 27 add -1 refcnt 4 actref 3 _state_del_locked + 604 in section .text of /lib/libganesha_nfsd.so.5.3.2
 idx 28 add -1 refcnt 3 actref 2 nfs4_op_close + 721 in section .text of /lib/libganesha_nfsd.so.5.3.2
 idx 29 add -1 refcnt 2 actref 1 compound_data_Free + 98 in section .text of /lib/libganesha_nfsd.so.5.3.2

**Comment by ffilz** [comment]
Actually it does help. That +1 refcnt is the sentinel ref. That it tracks one higher from the start indicates there is an upfront refcnt/actref that is not ever released. Also with the latest patchset, every ref other than the sentinel is ALSO an actref, so that is showing all the ref/unrefs from start to end.

I think mdcache_alloc_and_check_handle  is getting an extra refcnt/actref. I did hem and haw between mdcache_lru_get returning refcnt 2/actref 1 and refcnt 1/acrref 0 and mdcache_alloc_and_check_handle  getting an additional ref or not. I think I left the code out of sync...  Should be an easy fix.

**Comment by ffilz** [comment]
Yup, that was it... Plain as day in the code...

New patch set: https://review.gerrithub.io/c/ffilz/nfs-ganesha/+/556554?usp=search

I re-ordered the patches so the LRU changes are at HEAD (to make it easier to update it if testing reveals any other issues).

**Comment by Steve-Lord** [comment]
Looks like this is doing the trick, cache will not go past the high water mark.

Steve

**Comment by ffilz** [comment]
With this update, I have also determined there is absolutely no reason to revisit the number of lanes from the fixed 17. Every single attempt to reap an entry (even within a single call to lru_reap_impl) starts on a new lane using an atomic counter. In fact, because of this, the number of lanes actually didn't even need to be a prime number. At this point, there will almost never be lock contention, and the only reasons lru_reap_impl won't find an entry the first time is if the lane is actually empty or it is racing with an activation of the LRU entry in the first lane checked.

**Comment by ffilz** [comment]
Closing this now.
